core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -578,10 +614,14 @@ class Partition(val topicPartition: TopicPartition,
    * will never be added to ISR.
    *
    * Returns true if the HW was incremented, and false otherwise.
+   *
    * Note There is no need to acquire the leaderIsrUpdate lock here
-   * since all callers of this private API acquire that lock
+   * since all callers of this private API acquire either ISR read or write lock.
+   * HW update is performed while holding the Partition object lock with
+   * locking order ISR lock -> Partition object lock. So this method should not
+   * acquire an ISR read or write lock.
    */
-  private def maybeIncrementLeaderHW(leaderReplica: Replica, curTime: Long = time.milliseconds): Boolean = {
+  private def maybeIncrementLeaderHW(leaderReplica: Replica, curTime: Long = time.milliseconds): Boolean = synchronized {
```

Comment:
Looking at trunk, it seems like we no longer need this synchronization. We basically do the same thing by relying on the lock in `Log`. As far as I can tell, this is the only thing we really need to change in this patch.

Reply:
@hachikuji Thanks for the review,  I have rebased on trunk and removed this synchronization.